docs/concepts/CLQL.md

```diff
@@ -253,31 +253,226 @@ Results under include statements appear as children of the parent of the corresp
 A fact with multiple children will match against elements of the code that have child1 *and* child2 *and* child3 etc. The `any_of` operator overrides the implicit "and". The following query finds all String methods that use basic loops:
 
 ```
-method(depth = any):
+common.method(depth = any):
   name: “String”
   any_of:
-    foreach_stmt
-    while_stmt
-    for_stmt
+    common.foreach_stmt
+    common.while_stmt
+    common.for_stmt
 ```
 <!-- TODO(blakemscurr) n_of-->
 
 <br />
 
+# Edge
+
+Facts in AST lexicons refer to nodes in an AST, and the parent/child relationship between facts refers to the parent/child relationship of nodes in the AST. Nodes can have other parent/child relationships that are orthogonal to AST, such as calls. These relationships can be queried with the `edge` keyword.
+
+The following query finds function calls at the top level of a file and follows the `calls` edge to their definition:
+
+```
+common.func_call:
+  edge("calls"):
+    common.func
+```
+
+<br />
+
+# Block
+
+Block statements encapsulate CLQL trees. These subtrees can be repeated with a single argument allowing succint repition of complex patterns. Branches of the subtree can be recombined at the end of the block allowing queries along multiple paths.
+
+## Linear
+
+The following query finds functions with triply nested if statements:
+
+```clql
+common.func:
+  block(repeat = 3):
+    common.if_stmt:
+      blockcontinue
```

Comment:
this needs a `blockend`

Reply:
I was thinking that `blockend` is only necessary if you want to have multiple `blockcontinue` statements, and continue them after the block.

Reply:
This is an implicit / explicit decision. Generally, I'm in the explicit camp, but we are tending towards implicit with CLQL.

What happens if we nest a second block within the first? Does that start a new top level loop or is it a nested loop? Explicit blockends would make it clear.

Having said that, at this stage, I would be persuaded by whichever is 1. the quickest path to implement and 2. most likely to be confusion / side effect free.

Reply:
I'm not sure if this is an implicit / explicit issue. I might have a different view of what `blockend` means. Perhaps we could replace it with `afterblock`. To me, `blockend` means "match from here once you've finished with the block" - and if there is only one `blockcontinue` it seems natural to put the "match from here" nodes underneath the `blockcontinue`. 

I think nested block statements would create a nested loop, I'm not sure how they could be interpreted otherwise. I give an example of nested blocks below (though I don't think we need to support them yet for implementation reasons), and it doesn't have blockends. This "follows the callgraph from `someFunc` and via functions with multiply nested for loops":
```
common.func:
  name == "someFunc"
  block(repeat = any):
    block(repeat = 2:):
     common.for_stmt:
       blockcontinue:
        common.func_call(depth = any):
          edge("calls"):
            common.func:
              blockcontinue
```

I don't know what the implementation/side effect differences are, but I think the two would be much the same.

Reply:
You wouldn't nest two blocks directly. What about:

```
block(repeat = any):
  common.func:
  common.var
  block(repeat = 2):
    common.if
    blockcontinue
      common.for
  ```